src/ikernel.ts

```diff
@@ -404,6 +417,17 @@ namespace IKernel {
      * The done handler for the kernel future.
      */
     onDone: () => void;
+
+    /**
+     * Register hook for IOPub messages.
+     * 
+     * @returns a disposable to unregister the hook.
+     * 
+     * #### Notes
+     * The most recently-registered hook is run first.
+     */
+    registerMessageHook(hook: (msg: KernelMessage.IIOPubMessage) => boolean): void;
```

Comment:
Right. That was my iteration before this one. The problem again is that the thing registering the hook then holds a reference to the hook list, at least, and naively would also hold a reference to the future itself. That's what I was hoping avoid. In the end, I decided that registering a hook on the future was probably a relatively rare thing, since you could just modify the message handler directly. I kept the disposable on the kernel to still have a nice api there, but the disposable returned there doesn't retain a reference to the future or hook list.